File: leetcode/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        d = {}
        length = len(s)
        max_substring = 0

        prev_index = 0
        i = 0

        while i < length:
            d[s[i]] = d.get(s[i], 0) + 1

            if d[s[i]] > 1:
                max_substring = max(max_substring, i - prev_index)

                while d[s[i]] > 1:
                    d[s[prev_index]] -= 1 if d[s[prev_index]] > 0 else 0
                    prev_index += 1
            i += 1

        return max(max_substring, i - prev_index)

File: leetcode/longest-substring-without-repeating-characters.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    int lengthOfLongestSubstring(string s)
    {
        auto size = s.size();
        int current_max = 0;

        for (int i = 0; i < size; ++i) {
            int window_max = 1;
            unordered_map<char, bool> map;

            map[s[i]] = true;

            for (int j = i + 1; j < size; ++j) {
                if (map.find(s[j]) != map.end())
                    break;

                map[s[j]] = true;
                ++window_max;
            }
            current_max = max(window_max, current_max);
        }

        return current_max;
    }
};
